Implement Balanced Resource Allocation (BRA) algorithm as a PriorityFunction in scheduler package.

HaiyangDING opened the pull request:
This PR implements issue #5783.
- Add a new priority function: BalancedResourceAllocation
- BalancedResourceAllocation(BRA) rates a host based **ONLY** on how balanced its different types of resource are allocated if the pod would be scheduled on that host (metric: requested resource over capacity in terms of percentage for cpu and memory) . Score scales from 0 to 10 and 10 means the two percentage metrics are well very close, i.e. well balanced.
- BRA PriorityFunction should **NEVER** be used alone and **MUST** be used collaboratively with LeastRequestedPriority function.
- In some research articles, this method is referred to as Balanced Resource Utilization, and we choose to call it Balanced Resource Allocation for better readability.
### Why BRA

In the current PriorityFunction libraries, the only function considering the resource utilization is the LeastRequestedPriority function which would lead to imbalanced resource allocation in some case. For instance,  given two hosts with the same capacity {cpu: 10000, mem: 100000}, the current occupied resource is shown below:

<table>
    <tr>
        <td>host_name</td><td>cpu occupied</td><td>memory occupied</td>
    </tr>
    <tr>
        <td>host1</td><td>3000</td><td>40000</td>
    </tr>
    <tr>
        <td>host2</td><td>6000</td><td>10000</td>
    </tr>
</table>


when a pod with {cpu: 2000, mem: 10000}, the LeastRequestedPriority would score the two hosts as:
- host1:
  
  ```
  cpuScore := (10000 - 3000 - 2000) / 10000 * 10 = 5
  memScore := (100000 - 40000 - 10000) / 100000 * 10 = 5
  score := (5 + 5) / 2 = 5
  ```
- host2:
  
  ```
  cpuScore := (10000 - 6000 - 2000) / 10000 * 10 = 2
  memScore := (100000 - 10000 - 10000) / 100000 * 10 = 8
  score := (8 + 2) / 2 = 5
  ```

one can see that `host1.score` equals to `host2.score`, in which case the algorithm would select either host1 or host2 based on a random decision. However, it is clear that if host2 is chosen the resource usage rate would be imbalanced with 80% cpu occupied and only 20% memory used, leaving 80% as "resource fragment". 

Balanced Resource Allocation algorithm is here to address such problem. The rational underlying the algorithm is, for each host, evaluating the usage percentage of the resources (cpu and memoy in our case) assuming the pending pod already scheduled on that host and rating the host based on how balanced the concerned resources are used. That is to say, host with similar cpu usage percentage and memory usage percentage would be preferred over those without.

As for the example above, host1 would be preferred over host2 as host1 would end up with a more balanced resource usage rate.
### Details of BRA

The score given by BRA is scaled from 0-10 with 10 being preferred. This goes with other PriorityFunctions. Note that BRA implemented here considers only the percentage of requested resource over capacity without taking into account the actual free resources on the host. Thus, BRA PriorityFunction should **NEVER** be used alone and **MUST** be used collaboratively with LeastRequestedPriority function.

Here are the steps showing the score is calculated:

step1: calculate the cpuFraction and memoryFraction

```
            fractionCapacity:= Requested / Capacity * 100%
```

step2: calculate the difference

```
            diff := abs(cpuFraction - memoryFraction)
```

step3: scale the difference to 0~10 and subtract by 10 to get the score

```
            score := 10 - diff*10
```
### PS
- PS1: If `requested >= capacity`, score is set to 0 directly so that host with such high resource usage would not be preferred.
- PS3: For the example mentioned above, the final score would be: 
  
  ```
  host1.score = 10
  host2.score = 4
  ```
  
  so host1 is preferred by BRA function.

Commit:
Implement BRA algorithm as a new priority function in scheduler

Balanced Resource Allocation policy can now be enabled to so that
host(s) with balanced resource usage would be preferred. The score
given by BRA also scales from 0 to 10 with 10 representing that the
resource usage is well balanced.

erictune commented:
I'd rather we don't use the word "Usage" for  "Requested  divided by Capacity", unless there is some existing standard for naming these things, which I'm unaware of.

To me, "Usage" is an absolute quantity, not a ratio.  And, without qualifiers, it is a measure made by the operating system, not kubernetes pod requests.  See for example, the Linux kernel cpuacct.usage cgroup file.

Call this PctOfCapRequested or something.

erictune commented:
By the same logic, please don't call this Utilization.  IIUC, when people talk about utilization in a datacenter, without any qualifiers, they mean usage divided by capacity, not request divided by capacity.

Maybe call this Balanced Node Fullness, or something.

HaiyangDING commented:
Hi, @erictune 

Thank you for your comments.

I agree that using the word "Usage" for "Requested divided by Capacity" is not proper, and they will be replaced by "PctOfCapRequested" in both code and the PR message.

Regarding the word "Utilization" in Balanced Resource Utilization, this word is actually used in 2 research articles, i.e.

1. Xin Li et al, Balancing Resource Utilization for Continuous Virtual Machine Requests in Clouds, Sixth International Conference on Innovative Mobile and Internet Services in Ubiquitous Computing, 2012 
2. Wei Huang et al, An Energy Efficient Virtual Machine Placement Algorithm with Balanced Resource Utilization, Seventh International Conference on Innovative Mobile and Internet Services in Ubiquitous Computin, 2013

However, if you feel that this word is confusing in this scenario, I plan to change it to Balanced Resource Allocation. What do you think?

HaiyangDING commented:
Need some time to fix the check failure.

bgrant0607 commented:
cc @davidopp @abhgupta

HaiyangDING commented:
Hi, folks

My last commit has failed the continous-integration check and the reason is (as given by the details from https://travis-ci.org/GoogleCloudPlatform/kubernetes/jobs/56587523):

FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/controller	0.280s

However, I can not reproduce this error in my local environment, i.e. `./hacktest-go.sh` end up with all `ok`s. 

I am confused about this right now, so it would be great if someone could give any advice on this. Big thanks in advance!

abhgupta commented:
@HaiyangDING Seems to be an unrelated error due to a timeout - I would just retry the test. One way to kick off the tests again is to just modify the commit.

bgrant0607 commented:
I'll restart the tests.

erictune commented:
@HaiyangDING 
How about call it BalancedResourceAllocation, as you suggested, in the code and you can mention in the docstring that it is based on "Balanced Resource Utilization" research.

HaiyangDING commented:
@erictune 

Sounds like a great idea. Will implement sooner.

HaiyangDING commented:
@erictune 

PTAL, I have fixed the code and the description.

davidopp commented:
@erictune I assigned this to you just now. If you are too busy, I think @lavalamp said his PR review load is pretty light.

lavalamp commented on pkg/scheduler/priorities.go:
```diff
@@ -132,3 +134,92 @@ func (n *NodeLabelPrioritizer) CalculateNodeLabelPriority(pod api.Pod, podLister
+	capacityMilliCPU := node.Status.Capacity.Cpu().MilliValue()
+	capacityMemory := node.Status.Capacity.Memory().Value()
+
+	cpuPctOfCapRequested := calculatePctOfCapRequested(totalMilliCPU, capacityMilliCPU, node.Name)
```
suggest cpuFraction as the variable name. (and memoryFraction, averageFration below)

lavalamp replied:
Sorry for disagreeing with previous reviewer :(

lavalamp commented on pkg/scheduler/priorities.go:
```diff
@@ -132,3 +134,92 @@ func (n *NodeLabelPrioritizer) CalculateNodeLabelPriority(pod api.Pod, podLister
+	}
+}
+
+func calculatePctOfCapRequested(requested, capacity int64, node string) float64 {
```
How about fractionOfCapacity for a name, or capacityFraction?

(Reasoning: all functions "calculate" things, so there's no need to state so in the name, and fraction is clearer than percent, which could leave the reader thinking you've multiplied the result by 100.)

lavalamp commented on pkg/scheduler/priorities.go:
```diff
@@ -132,3 +134,92 @@ func (n *NodeLabelPrioritizer) CalculateNodeLabelPriority(pod api.Pod, podLister
 	return result, nil
 }
+
+// BalancedResourceAllocation is favors nodes with balanced resource usage rate.
```
Did you get this algorithm from this paper? "An Energy Efficient Virtual Machine Placement
Algorithm with Balanced Resource Utilization" http://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=6603690

Please document if so-- it explains how to extend the math to further dimensions.

lavalamp replied:
This is my last request, I think-- document sources so people know what they need to read if this code needs to be updated.

lavalamp commented on pkg/scheduler/priorities.go:
```diff
@@ -132,3 +134,92 @@ func (n *NodeLabelPrioritizer) CalculateNodeLabelPriority(pod api.Pod, podLister
+	cpuDiff := cpuPctOfCapRequested - avergePctOfCapRequested
+	memoryDiff := memoryPctOfCapRequested - avergePctOfCapRequested
+	diff := math.Sqrt(math.Pow(cpuDiff, 2) + math.Pow(memoryDiff, 2))
+	// closer percentage metrics lead to higher score with 10 being the highest
```
I'm not really sure about this scaling mechanism of mathematical inverse here, it seems ad-hoc?

Compare: https://www.wolframalpha.com/input/?i=graph+sqrt%28x*x+%2B+y*y%29

vs: https://www.wolframalpha.com/input/?i=graph+1%2Fsqrt%28x*x+%2B+y*y%29

The former looks much more gentle than the latter, no?

What do you think about:
a) compute the upper and lower bound of diff
b) scale to the range 0-10
c) then subtract this from 10 instead of doing the inverse.

Or maybe I'm missing something important?

HaiyangDING replied:
Hi @lavalamp

I appreciate your comments here very much and I am going to change them accordingly. About the proposal of the scaling mechanism, here are some of my thoughts:

1. I agree with you that there are many more gentle ways to deal with difference other than that in the current patch. 
2. As long as the score is capable to prefer similar fraction of capability, I don't think we could miss anything essential here. 
3. Given that the LeastRequestedPriority function is also a "gentle" method, I think we'd better off doing the same here.

However, I am not sure that I have fully understood your proposal, and to the best of my understanding(should be very similar to yours), here is what I am going to change it:

1. consider the difference between cpuFraction and memoryFraction directly, the lower and upper bound are (-1.0) ~ (1.0)
2. use absolute value here and scale to 0-10: |cpuFraction-memoryFraction| * 10
3. subtract from 10 to have the score whose graph is like this: https://www.wolframalpha.com/input/?i=graph+10-%7Cx-y%7C*10%2C+x+in+%280%2C1%29%2C+y+in+%280%2C+1%29, which is gentle and intuitive enough to show the essence of this priority function.

note: the computation could start from calculating the averageFraction as well, but it would end up with the same solution as above.

I am going to change scale mechanism this way. What do you think?

PS: I will squash them as well.

lavalamp replied:
I think your proposed mechanism is better, but I think best would be if you kept the euclidean distance metric (for maximum adherence to the literature). Your step 1 switches to a Manhattan distance metric. So retain the sqrt(cpu^2 + mem^2) step.

lavalamp replied:
Of course this might be moot since the thing we do with the priority definitely doesn't match any literature...

HaiyangDING replied:
Hi, maybe I did not make myself very clear...... I googled Manhatten distance and I think what I am suggesting is not.

sqrt(cpu^2 + mem^2) itself cannot reflect how far cpu and mem are from each other, so one would eventually have to use sqrt((cpu-average)^2+(mem-average)^2) to reflect the distance (or without sqrt). Given that average=(cpu + mem)/2, $sqrt((cpu-average)^2+(mem-average)^2)$ and $abs(cpu - mem)$ are in fact both euclidean distance in difference scale, i.e.:

sqrt((cpu-average)^2+(mem-average)^2) = 
sqrt((cpu-(cpu+mem)/2)^2 + (mem-(cpu+mem)/2)^2) = 
sqrt(((cpu-mem)/2)^2 + ((mem-cpu)/2)^2) = 
sqrt((cpu-mem)^2/4) = sqrt((cpu-mem)^2)/2 = 
abs(cpu-mem) / 2

I have submitted a new patch to implement the my ideas (although it needs rebase to pass ci test). By looking at the test cases, I think the function gives reasonble scores for different degree of resource balance.

All in all, I can use sqrt instead of absolute vaule, but as I explained above, they end up the same and what matters only is how to scale them from 0 - 10. Of course, using sqrt may be more intuitive for one to realize that we are considering the distance between resource types.

What do you think?

lavalamp replied:
It looks like you're correct. I end up with (sqrt(2)/2)*abs(cpu-mem) when I do the math, but we're scaling it, so it doesn't make any difference. Thanks for bearing with me!

HaiyangDING replied:
Oh, my pleasure. Speaking of "thanks", it should be me thanking you for all the patience during improving this PR:)

lavalamp commented on pkg/scheduler/priorities.go:
```diff
@@ -132,3 +134,92 @@ func (n *NodeLabelPrioritizer) CalculateNodeLabelPriority(pod api.Pod, podLister
+	// isEqual() is used to eliminate the risk of value inaccuracy.
+	// Higher threhold of the acccuracy parameter is beyond the scope of BRU algorithm; as said in the introduction
+	// this BRU PriorityFunction should NEVER be used alone.
+	if isEqual(cpuPctOfCapRequested, float64(1)) || isEqual(memoryPctOfCapRequested, float64(1)) {
```
the float64() conversion isn't needed-- constants aren't typed in go, so the language will do the right thing.

lavalamp commented on pkg/scheduler/priorities.go:
```diff
@@ -132,3 +134,92 @@ func (n *NodeLabelPrioritizer) CalculateNodeLabelPriority(pod api.Pod, podLister
+	} else {
+		score = int(1 / diff)
+	}
+	// if Requesetd/Capacity exceeds 100%, the PctOfCapRequested is set to 1 (100%) and the host should never be preferred.
```
Why not just directly check for request >= capacity above (before the call to calculatePctOfRequested) and return early if so?

lavalamp commented on pkg/scheduler/priorities.go:
```diff
@@ -132,3 +134,92 @@ func (n *NodeLabelPrioritizer) CalculateNodeLabelPriority(pod api.Pod, podLister
+		return float64(1)
+	}
+	if requested > capacity {
+		glog.Errorf("Combined requested resources from existing pods exceeds capacity on minion: %s", node)
```
Don't log this, it's not an error. Recall that you added in the prospective pod's resource usage-- presumably that's what caused the overage.

lavalamp commented:
Thanks for this change! I have a few comments. They're mostly minor.

Please squash your commits, in case I forget to ask later.

HaiyangDING commented:
@lavalamp

Hi, thank you for the kind comments. I have just returned from a small vacation. Will be working on it sooner.

HaiyangDING commented:
New patch submitted at my home where tests are unable to perform, I will fix them tomorrow. Sorry for the trouble.

bgrant0607 commented on pkg/scheduler/priorities_test.go:
```diff
@@ -364,3 +364,232 @@ func TestNewNodeLabelPriority(t *testing.T) {
+		"baz": "blah",
+	}
+	machine1Status := api.PodStatus{
+		Host: "machine1",
```
Needs to be rebased. This field is gone.

HaiyangDING replied:
Thank you, I will fix them sooner.

HaiyangDING commented:
How to deal with Shippable failures? I can not see the details.

lavalamp commented:
Shippable is an unrelated flake. Last request before LGTM: document sources, and squash commits again. Thanks!

HaiyangDING commented:
I confirm that all contributors to this PR are OK with these commits being contributed to Google.

As a matter of fact, the commits are all made by me (dingh, HaiyangDING, dinghaiyang@huawei.com) but throught different PCs

PS: I will squash all the commits as I did in https://github.com/HaiyangDING/kubernetes/commit/15b91cd932531b1d41293578832344191d20fee4 later

HaiyangDING commented:
I confirm that all contributors to this PR are OK with these commits being contributed to Google.

As a matter of fact, the commits are all made by me (dingh, HaiyangDING, dinghaiyang@huawei.com<mailto:dinghaiyang@huawei.com>) but throught different PCs

丁海洋Haiyang DING,
华为技术有限公司.
Technology Introduction & Standard and Patent Dept,
Carrier Software Business Unit,
Huawei Technologies.
地址：南京市雨花区软件大道 101 号.
Address: Huawei Nanjing R&D Center
101 Software Avenue,Yuhuatai District, Nanjing 210012
Jiangsu, P.R.China
Tel: 0086-025-56620756
Email: dinghaiyang@huawei.com<mailto:huruifeng@huawei.com>
http://www.huawei.com

发件人: googlebot [mailto:notifications@github.com]
发送时间: 2015年4月9日 15:34
收件人: GoogleCloudPlatform/kubernetes
抄送: dinghaiyang
主题: Re: [kubernetes] Implement Balanced Resource Allocation (BRA) algorithm as a PriorityFunction in scheduler package. (#6150)


We found a Contributor License Agreement for you (the sender of this pull request) and all commit authors, but as best as we can tell these commits were authored by someone else. If that's the case, please add them to this pull request and have them confirm that they're okay with these commits being contributed to Google. If we're mistaken and you did author these commits, just reply here to confirm.

—
Reply to this email directly or view it on GitHub<https://github.com/GoogleCloudPlatform/kubernetes/pull/6150#issuecomment-91137643>.

HaiyangDING commented:
@lavalamp 

I think I have squashed all the commits, but it seems I have to do `git pull` + `git push` after `git rebase -i` which brings up the last commit (Merge branch 'bru' of https://github.com/HaiyangDING/kubernetes into bru 	952c182). Details are at below.

I wonder if it is OK. Please bear me with this, I am new to squashing and let me know if anything should be done.

------------------- details of last steps of `git rebase` --------------

openstack@network:~/dingh/kubernetes$ git rebase --continue
[detached HEAD 4507b9f] Implement BRA algorithm as a new priority function in scheduler
 2 files changed, 309 insertions(+)
[detached HEAD 2bb7eeb] Implement BRA algorithm as a new priority function in scheduler
 2 files changed, 309 insertions(+)
Successfully rebased and updated refs/heads/bru.
openstack@network:~/dingh/kubernetes$ git push
Username for 'https://github.com': dinghaiyang@huawei.com
Password for 'https://dinghaiyang@huawei.com@github.com': 
To https://github.com/HaiyangDING/kubernetes.git
 ! [rejected]        bru -> bru (non-fast-forward)
error: failed to push some refs to 'https://github.com/HaiyangDING/kubernetes.git'
To prevent you from losing history, non-fast-forward updates were rejected
Merge the remote changes (e.g. 'git pull') before pushing again.  See the
'Note about fast-forwards' section of 'git push --help' for details.
openstack@network:~/dingh/kubernetes$ git pull
Merge made by the 'recursive' strategy.
openstack@network:~/dingh/kubernetes$ git push
Username for 'https://github.com': dinghaiyang@huawei.com
Password for 'https://dinghaiyang@huawei.com@github.com': 
To https://github.com/HaiyangDING/kubernetes.git
   5387fab..952c182  bru -> bru
openstack@network:~/dingh/kubernetes$ git push --help

lavalamp commented:
@HaiyangDING use fetch + rebase instead of pull + merge to prevent that.

HaiyangDING commented:
@lavalamp 

PTAL, finally it works... thanks

lavalamp commented:
LGTM, thank you!
